install.sh

```diff
@@ -0,0 +1,459 @@
+}
+version_lt() {
+  [[ "${1%.*}" -lt "${2%.*}" ]] || [[ "${1%.*}" -eq "${2%.*}" && "${1#*.}" -lt "${2#*.}" ]]
+}
+
+should_install_command_line_tools() {
+  if version_gt "$macos_version" "10.13"; then
+    ! [[ -e "/Library/Developer/CommandLineTools/usr/bin/git" ]]
+  else
+    ! [[ -e "/Library/Developer/CommandLineTools/usr/bin/git" ]] ||
+      ! [[ -e "/usr/include/iconv.h" ]]
+  fi
+}
+
+get_permission() {
+  stat -f "%A" "$1"
+}
+
+user_only_chmod() {
+  [[ -d "$1" ]] && [[ "$(get_permission "$1")" != "755" ]]
+}
+
+exists_but_not_writable() {
+  [[ -e "$1" ]] && ! [[ -r "$1" && -w "$1" && -x "$1" ]]
+}
+
+get_owner() {
+  stat -f "%u" "$1"
+}
+
+file_not_owned() {
+  [[ "$(get_owner "$1")" != "$(id -u)" ]]
+}
+
+get_group() {
+  stat -f "%g" "$1"
+}
+
+file_not_grpowned() {
+  [[ " $(id -G "$USER") " != *" $(get_group "$1") "*  ]]
+}
+
+# USER isn't always set so provide a fall back for the installer and subprocesses.
+if [[ -z "$USER" ]]; then
+  USER="$(chomp "$(id -un)")"
+  export USER
+fi
+
+# Invalidate sudo timestamp before exiting (if it wasn't active before).
+if ! /usr/bin/sudo -n -v 2>/dev/null; then
+  trap '/usr/bin/sudo -k' EXIT
+fi
+
+# Things can fail later if `pwd` doesn't exist.
+# Also sudo prints a warning message for no good reason
+cd "/usr" || exit 1
+
+####################################################################### script
+if [[ "$OSTYPE" == "linux-gnu" ]]; then
+  abort "$(cat <<'EOABORT'
+  To install Homebrew on Linux, paste at a terminal prompt:
+    sh -c "$(curl -fsSL https://raw.githubusercontent.com/Linuxbrew/install/master/install.sh)"
+EOABORT
+)"
+elif version_lt "$macos_version" "10.7"; then
+  abort "$(cat <<EOABORT
+Your Mac OS X version is too old. See:
+  ${tty_underline}https://github.com/mistydemeo/tigerbrew${tty_reset}
+EOABORT
+)"
+elif version_lt "$macos_version" "10.9"; then
+  abort "Your OS X version is too old"
+elif [[ "$UID" == "0" ]]; then
+  abort "Don't run this as root!"
+elif ! [[ "$(dsmemberutil checkmembership -U "$USER" -G admin)" = *"user is a member"* ]]; then
+  abort "This script requires the user $USER to be an Administrator."
+elif [[ -d "$HOMEBREW_PREFIX" && ! -x "$HOMEBREW_PREFIX" ]]; then
+  abort "$(cat <<EOABORT
+The Homebrew prefix, ${HOMEBREW_PREFIX}, exists but is not searchable. If this is
+not intentional, please restore the default permissions and try running the
+installer again:
+    sudo chmod 775 ${HOMEBREW_PREFIX}
+EOABORT
+)"
+elif version_gt "$macos_version" "$MACOS_LATEST_SUPPORTED" || \
+  version_lt "$macos_version" "$MACOS_OLDEST_SUPPORTED"; then
+  who="We"
+  what=""
+  if version_gt "$macos_version" "$MACOS_LATEST_SUPPORTED"; then
+    what="pre-release version"
+  else
+    who+=" (and Apple)"
+    what="old version"
+  fi
+  ohai "You are using macOS ${macos_version}."
+  ohai "${who} do not provide support for this ${what}."
+
+  echo "$(cat <<EOS
```

Comment:
Again, I don’t get the use of `cat`, as `echo` alone would suffice. As we can make it look nicer by keeping the indentation and removing it on output, like so:

```bash
echo "
  Some message
  With ${variables}
" sed -E 's/^ {2}//'
```

Replacing the `2` with however many spaces we have at the start of the line.

Reply:
I personally prefer simpler and less magical (no undenting).
```sh
cat <<EOS
Hello, world!
EOS
```

Reply:
I disagree that the `sed` call is either complex or magical, especially when you consider the script as a whole, but I won’t bikeshed it. Not using `echo` and `cat` at the same time should be considered, though.

There’s also

```bash
cat <<-EOS
	Hello
	World
EOS
```

The `<<-` removes the indentation (not as good as Ruby’s `<<~`), but it only works with tabs.

Reply:
Undenting is also probably not worth the complexity, agreed.

Reply:
Or we could just `echo` every line separately, which also aligns nicely with `ohai`, as seen below.